js/src/BTCUtils.js

```diff
@@ -1,640 +1,772 @@
-   * @notice                Extracts the timestamp from a block header
-   * @dev                   Time is not 100% reliable
-   * @param {Uint8Array}    header The header
-   * @returns {BigInt}      The timestamp (uint)
-   */
-  extractTimestamp: header => utils.bytesToUint(
-    module.exports.reverseEndianness(module.exports.extractTimestampLE(header))
-  ),
-
-  /**
-   * @notice                Extracts the expected difficulty from a block header
-   * @dev                   Does NOT verify the work
-   * @param {Uint8Array}    header The header
-   * @returns {BigInt}      The difficulty as an integer
-   */
-  extractDifficulty: header => module.exports.calculateDifficulty(
-    module.exports.extractTarget(header)
-  ),
-
-  /**
-   * @notice                Concatenates and hashes two inputs for merkle proving
-   * @param {Uint8Array}    a The first hash
-   * @param {Uint8Array}    b The second hash
-   * @returns {Uint8Array}  The double-sha256 of the concatenated hashes
-   */
-  hash256MerkleStep: (a, b) => module.exports.hash256(utils.concatUint8Arrays(a, b)),
-
-  /**
-   * @notice                Verifies a Bitcoin-style merkle tree
-   * @dev                   Leaves are 1-indexed.
-   * @param {Uint8Array}     proof The proof. Tightly packed LE sha256 hashes.
-   *                        The last hash is the root
-   * @param {Number}        index The index of the leaf
-   * @returns {Boolean}     True if the proof is value, else false
-   */
-  verifyHash256Merkle: (proof, index) => {
-    let idx = index;
-    const proofLength = proof.length;
-
-    // Not an even number of hashes
-    if (proofLength % 32 !== 0) {
-      return false;
-    }
+  // Returns false if we're not exactly at the end
+  return offset === vLength;
+}
 
-    // Special case for coinbase-only blocks
-    if (proofLength === 32) {
-      return true;
-    }
+/**
+ *
+ * Checks that the vout passed up is properly formatted
+ *
+ * @dev                   Consider a vin with a valid vout in its scriptsig
+ * @param {Uint8Array}    vout Raw bytes length-prefixed output vector
+ * @returns {Boolean}     True if it represents a validly formatted vout
+ */
+export function validateVout(vout) {
+  let offset = BigInt(1);
+  const vLength = BigInt(vout.length);
+  const [nOuts] = vout;
+
+  // Not valid if it says there are too many or no inputs
+  if (nOuts >= 0xfd || nOuts === 0) {
+    return false;
+  }
 
-    // Should never occur
-    if (proofLength === 64) {
-      return false;
-    }
+  for (let i = 0; i < nOuts; i += 1) {
+    // Grab the next input and determine its length.
+    // Increase the offset by that much
+    offset += determineOutputLength(utils.safeSlice(vout, offset));
+  }
 
-    const root = utils.safeSlice(proof, (proofLength - 32), proofLength);
-    let current = utils.safeSlice(proof, 0, 32);
-    const length = (proofLength / 32) - 1;
+  // Returns false if we jump past the end
+  if (offset > vLength) {
+    return false;
+  }
```

Comment:
Doesn't this do the same thing as ```return offset === vLength;``` on the line below?  Or maybe it's supposed to go inside the for loop like in `validateVin`?

Reply:
Good catch, it was supposed to be inside